Windows CI builds

bhaskarvk opened the pull request:
We can't leave the `windows` users guessing :-)

The build works ( [full build log](https://ci.appveyor.com/project/bhaskarvk/reticulate)), but I do see some warnings. Also 9 out of the 50 tests were skipped.

```
* checking compiled code ... NOTE
Warning in read_symbols_from_dll(so, rarch) :
  this requires 'objdump.exe' to be on the PATH
Warning in read_symbols_from_dll(so, rarch) :
  this requires 'objdump.exe' to be on the PATH
Warning in read_symbols_from_dll(so, rarch) :
  this requires 'objdump.exe' to be on the PATH
File 'reticulate/libs/i386/reticulate.dll':
  Found no calls to: 'R_registerRoutines', 'R_useDynamicSymbols'

It is good practice to register native routines and to disable symbol
search.

See 'Writing portable packages' in the 'Writing R Extensions' manual.
* checking examples ... OK
* checking for unstated dependencies in 'tests' ... OK
* checking tests ... OK
  Running 'testthat.R' [11s]
```

jjallaire commented:
It looks like you do need to ensure that objdump.exe is on the PATH as those calls are definitely made as required: https://github.com/rstudio/reticulate/blob/3e08b34b711becaaf868e39d2d9e08b2c9d1bac4/src/RcppExports.cpp#L500-L503

My guess is that the tests are skipped because NumPy isn't available. Is there a way to get it installed?

bhaskarvk commented:
Yes there is, I will add the steps to the appveyour.xml

bhaskarvk commented:
I am still testing this, so you'll see some commits as I tune the file a bit. I'll give you the final green light when I've verified everything.

bhaskarvk commented:
I added steps to install numpy and also test against python 2.7, 3.4, 3.5, 3.6.
The `objdjump` warnings are gone after I added MinGW to the Path, now just one warning

```
File 'reticulate/libs/i386/reticulate.dll':
  Found '_ZSt4cerr', possibly from 'std::cerr' (C++)
    Object: 'python.o'
Compiled code should not call entry points which might terminate R nor
write to stdout/stderr instead of to the console, nor the system RNG.
See 'Writing portable packages' in the 'Writing R Extensions' manual.
* checking examples ... OK
* checking for unstated dependencies in 'tests' ... OK
* checking tests ...
```

I do have numpy installed now, but still 9 tests are being skipped. Unfortunately testthat doesn't provide much detail beyond that, like which tests are skipped and why, so I added `py_discover_config` and `py_config` calls in the test R file to debug, and it seems reticulate can indeed find numpy correctly.

```
R version 3.4.1 Patched (2017-07-09 r72907) -- "Single Candle"

> library(testthat)
> library(reticulate)
> 
> py_discover_config("numpy")
python:         C:\Python27/python.exe
libpython:      python27.dll
pythonhome:     C:\Python27
version:        2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 20:42:59) [MSC v.1500 32 bit (Intel)]
Architecture:   32bit
numpy:          C:\Python27\lib\site-packages\numpy
numpy_version:  1.13.1
numpy:          C:\Python27\lib\site-packages\numpy

python versions found: 
 C:\Python27/python.exe
 c:\Python27\python.exe
 C:\PY7FEF~1\\python.exe
 C:\Python35\\python.exe
 C:\PY2ABB~1\\python.exe
 C:\Python36\\python.exe
 C:\Python26\\python.exe
 C:\Python27\\python.exe
 C:\Python33\\python.exe
 C:\Python34\\python.exe
 C:\MIB14C~1\python.exe
 C:\MI8349~1\python.exe
 C:\MINICO~2\python.exe
 C:\MI0C47~1\python.exe
> py_config()
python:         C:\Python27/python.exe
libpython:      python27.dll
pythonhome:     C:\Python27
version:        2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 20:42:59) [MSC v.1500 32 bit (Intel)]
Architecture:   32bit
numpy:          C:\Python27\lib\site-packages\numpy
numpy_version:  1.13.1

python versions found: 
 C:\Python27/python.exe
 c:\Python27\python.exe
 C:\PY7FEF~1\\python.exe
 C:\Python35\\python.exe
 C:\PY2ABB~1\\python.exe
 C:\Python36\\python.exe
 C:\Python26\\python.exe
 C:\Python27\\python.exe
 C:\Python33\\python.exe
 C:\Python34\\python.exe
 C:\MIB14C~1\python.exe
 C:\MI8349~1\python.exe
 C:\MINICO~2\python.exe
 C:\MI0C47~1\python.exe
> 
> test_check("reticulate")
testthat results ================================================================
OK: 50 SKIPPED: 9 FAILED: 0
> 
> proc.time()
   user  system elapsed 
   1.34    0.12    3.37 
```

Here's the full test https://ci.appveyor.com/project/bhaskarvk/reticulate

I can take off the debug statements if you like, but I kinda like them for quick debugging. Either way your call. I don't think I can fix anything beyond this, but I still feel it's worthwhile to test against windows.

jjallaire commented on tests/testthat.R:
```diff
@@ -1,4 +1,7 @@
 library(testthat)
 library(reticulate)
 
+py_discover_config("numpy")
```
You can't just call `py_config` unqualified as it will fail on systems where there is no compatible version of Python (e.g. some CRAN servers). That's why we use `skip_if_no_python()`

jjallaire commented:
We have some calls to `skip_on_cran()`. Perhaps the `NOT_CRAN` environment
variable isn't defined on Appveyor?

On Thu, Jul 13, 2017 at 8:36 AM, Bhaskar V. Karambelkar <
notifications@github.com> wrote:

> I added steps to install numpy and also test against python 2.7, 3.4, 3.5,
> 3.6.
> The objdjump warnings are gone after I added MinGW to the Path, now just
> one warning
>
> File 'reticulate/libs/i386/reticulate.dll':
>   Found '_ZSt4cerr', possibly from 'std::cerr' (C++)
>     Object: 'python.o'
> Compiled code should not call entry points which might terminate R nor
> write to stdout/stderr instead of to the console, nor the system RNG.
> See 'Writing portable packages' in the 'Writing R Extensions' manual.
> * checking examples ... OK
> * checking for unstated dependencies in 'tests' ... OK
> * checking tests ...
>
> I do have numpy installed now, but still 9 tests are being skipped.
> Unfortunately testthat doesn't provide much detail beyond that, like which
> tests are skipped and why, so I added py_discover_config and py_config
> calls in the test R file to debug, and it seems reticulate can indeed find
> numpy correctly.
>
> R version 3.4.1 Patched (2017-07-09 r72907) -- "Single Candle"
>
> > library(testthat)
> > library(reticulate)
> >
> > py_discover_config("numpy")
> python:         C:\Python27/python.exe
> libpython:      python27.dll
> pythonhome:     C:\Python27
> version:        2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 20:42:59) [MSC v.1500 32 bit (Intel)]
> Architecture:   32bit
> numpy:          C:\Python27\lib\site-packages\numpy
> numpy_version:  1.13.1
> numpy:          C:\Python27\lib\site-packages\numpy
>
> python versions found:
>  C:\Python27/python.exe
>  c:\Python27\python.exe
>  C:\PY7FEF~1\\python.exe
>  C:\Python35\\python.exe
>  C:\PY2ABB~1\\python.exe
>  C:\Python36\\python.exe
>  C:\Python26\\python.exe
>  C:\Python27\\python.exe
>  C:\Python33\\python.exe
>  C:\Python34\\python.exe
>  C:\MIB14C~1\python.exe
>  C:\MI8349~1\python.exe
>  C:\MINICO~2\python.exe
>  C:\MI0C47~1\python.exe
> > py_config()
> python:         C:\Python27/python.exe
> libpython:      python27.dll
> pythonhome:     C:\Python27
> version:        2.7.13 (v2.7.13:a06454b1afa1, Dec 17 2016, 20:42:59) [MSC v.1500 32 bit (Intel)]
> Architecture:   32bit
> numpy:          C:\Python27\lib\site-packages\numpy
> numpy_version:  1.13.1
>
> python versions found:
>  C:\Python27/python.exe
>  c:\Python27\python.exe
>  C:\PY7FEF~1\\python.exe
>  C:\Python35\\python.exe
>  C:\PY2ABB~1\\python.exe
>  C:\Python36\\python.exe
>  C:\Python26\\python.exe
>  C:\Python27\\python.exe
>  C:\Python33\\python.exe
>  C:\Python34\\python.exe
>  C:\MIB14C~1\python.exe
>  C:\MI8349~1\python.exe
>  C:\MINICO~2\python.exe
>  C:\MI0C47~1\python.exe
> >
> > test_check("reticulate")
> testthat results ================================================================
> OK: 50 SKIPPED: 9 FAILED: 0
> >
> > proc.time()
>    user  system elapsed
>    1.34    0.12    3.37
>
> Here's the full test https://ci.appveyor.com/project/bhaskarvk/reticulate
>
> I can take off the debug statements if you like, but I kinda like them for
> quick debugging. Either way your call. I don't think I can fix anything
> beyond this, but I still feel it's worthwhile to test against windows.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rstudio/reticulate/pull/78#issuecomment-315064835>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGXxzlsoUSvilOJ-JPm4uq720g0Gcy8ks5sNg9RgaJpZM4OWfFM>
> .
>

bhaskarvk commented:
Took out the debug statements and added NOT_CRAN. Will report back in a few.

jjallaire commented:
I just made a change to use `PySys_WriteStderr` rather than `std::cerr` which should avoid the warning about accessing std::cerr

jjallaire commented:
I am going to remove the py_function_wrapper tests as they are (a) very hard to maintain as they deal with exact reproduction of text output; and (b) not really critical as they are just a utility function for mostly internal use.

bhaskarvk commented:
Sounds good. I'll circle back sometime in the afternoon.

jjallaire commented:
Okay, that's done on master so if you merge from master you should now get a clean run. Let me know once you've confirmed and I'll merge.

bhaskarvk commented:
All done! All 61 tests pass on  2.7 & 3.4-3.6. Good to go.
https://ci.appveyor.com/project/bhaskarvk/reticulate
You will need to setup the CI build on appveyor using the rstudio a/c and also add the badge in README.

jjallaire commented:
Thanks! I seem to be unable to add the reticulate repo on appveyor -- any
hints/tips you can share?

On Thu, Jul 13, 2017 at 10:54 AM, Bhaskar V. Karambelkar <
notifications@github.com> wrote:

> All done! All 61 tests pass on 2.7 & 3.4-3.6. Good to go.
> https://ci.appveyor.com/project/bhaskarvk/reticulate
> You will need to setup the CI build on appveyor using the rstudio a/c and
> also add the badge in README.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rstudio/reticulate/pull/78#issuecomment-315102430>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGXx2c1120-LUh5fWqQPAZni8merZT2ks5sNi-QgaJpZM4OWfFM>
> .
>

bhaskarvk commented:
Go to https://ci.appveyor.com/, and log in with your GH a/c . then it should prompt you to add a new project where you can select reticulate from the GH projects it lists.

jjallaire commented:
Unfortunately it doesn't list reticulate for me (although it lists dozens
of other projects!).



On Thu, Jul 13, 2017 at 1:43 PM, Bhaskar V. Karambelkar <
notifications@github.com> wrote:

> Go to https://ci.appveyor.com/, and log in with your GH a/c . then it
> should prompt you to add a new project where you can select reticulate from
> the GH projects it lists.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/rstudio/reticulate/pull/78#issuecomment-315150769>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGXx6o3WF-7BypbCnPJFiq1EMkrH85lks5sNldCgaJpZM4OWfFM>
> .
>

bhaskarvk commented:
Interesting. I will investigate.

jjallaire commented:
Okay, thanks!

On Thu, Jul 13, 2017 at 2:17 PM, Bhaskar V. Karambelkar <
notifications@github.com> wrote:

> Interesting. I will investigate.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/rstudio/reticulate/pull/78#issuecomment-315160031>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGXx_6CQxKdT_Y76z5E3OrLMPdSuif-ks5sNl8sgaJpZM4OWfFM>
> .
>

bhaskarvk commented:
https://www.appveyor.com/docs/team-setup/ has some steps outlined.
When I go to add a project I see my projects organized by my public projects followed by the projects I have access to by being part of an Org. May be double check your a/c page 
https://ci.appveyor.com/account
Also googling turned up this http://help.appveyor.com/discussions/problems/142-unable-to-create-a-project-from-github-org-account

jjallaire commented:
Okay, that was extremely helpful! It looks like in order for an RStudio repo to be added to appveyor @jimhester must be one of the collaborators on that repo (presumably b/c he's the one that set up the initial connection to the rstudio org on appveyor). I added him and all is well.

bhaskarvk commented:
I am getting very serious about CI builds on all 3 platforms. So I might give you guys PRs for other RStudio projects from time to time.

jjallaire commented:
Just added the appveyor status badge. Thanks again!!!!
